src/logging.c

```diff
@@ -104,47 +173,149 @@ void rcutils_logging_console_output_handler(
+  // dynamically allocate space.
+  output_buffer = static_output_buffer;
+  memset(static_output_buffer, '\0', sizeof(static_output_buffer));
+  size_t output_buffer_size = sizeof(static_output_buffer);
+  const char * str = g_rcutils_logging_output_format_string;
+  size_t size = strlen(g_rcutils_logging_output_format_string);
+
+  // Walk through the format string and expand tokens when they're encountered.
+  size_t n = 0;
+  size_t i = 0;
+  const char * token_expansion = NULL;
+  while (i < size) {
+    // Print everything up to the next token start delimiter.
+    size_t chars_to_start_delim = rcutils_find(str + i, token_start_delimiter);
+    size_t remaining_chars = size - i;
+    if (chars_to_start_delim > 0) {
+      if (chars_to_start_delim > remaining_chars) {
+        // No start delimiters found; don't allow printing more of than what's left if.
+        chars_to_start_delim = remaining_chars;
+      }
+      RCUTILS_LOGGING_ENSURE_LARGE_ENOUGH_BUFFER(
+        chars_to_start_delim, output_buffer_size, allocator, output_buffer, static_output_buffer)
+      memcpy(output_buffer + strlen(output_buffer), str + i, chars_to_start_delim);
+      i += chars_to_start_delim;
+      if (i >= size) {
+        break;
+      }
+    }
+    // We are at a token start delimiter: determine if there's a known token or not.
+    // Potential tokens can't possibly be longer than the format string itself.
+    char token[RCUTILS_LOGGING_MAX_OUTPUT_FORMAT_LEN];
+    // Look for a token end delimiter.
+    size_t chars_to_end_delim = rcutils_find(str + i, token_end_delimiter);
+    remaining_chars = size - i;
+    if (chars_to_end_delim > remaining_chars) {
+      // No end delimiters found in the remainder of the format string;
+      // there won't be any more tokens so shortcut the rest of the checking.
+      RCUTILS_LOGGING_ENSURE_LARGE_ENOUGH_BUFFER(
+        remaining_chars, output_buffer_size, allocator, output_buffer, static_output_buffer)
+      memcpy(output_buffer + strlen(output_buffer), str + i, remaining_chars);
+      break;
+    }
+    // Found what looks like a token; determine if it's recognized.
+    size_t token_len = chars_to_end_delim - 1;  // Not including delimiters.
+    memcpy(token, str + i + 1, token_len);  // Skip the start delimiter.
+    token[token_len] = '\0';
+    if (strcmp("severity", token) == 0) {
+      token_expansion = severity_string;
+    } else if (strcmp("name", token) == 0) {
+      token_expansion = name;
+    } else if (strcmp("message", token) == 0) {
+      token_expansion = message_buffer;
+    } else if (strcmp("function_name", token) == 0) {
+      token_expansion = location ? location->function_name : "\"\"";
+    } else if (strcmp("file_name", token) == 0) {
+      token_expansion = location ? location->file_name : "\"\"";
+    } else if (strcmp("line_number", token) == 0) {
+      if (location) {
+        char line_number_expansion[10];  // Allow 9 digits for the expansion (otherwise, truncate).
```

Comment:
I am certain this will be enough to cover line numbers. I wonder though if we need to limit it artificially. Could the code instead us `SIZE_MAX` and derive the maximum number of digits from it? Then there is no need to truncate and the size of the buffer should be only slightly bigger than currently. But it will allow to get rid of the case to handle the truncation.

Reply:
the truncation is handled by `rcutils_snprintf` and FWIU we'll need to use that function regardless. Because of that, while I like the idea of determining the max number of digits, I don't think it will gain us any simplification in the code, so to me it doesn't seem it'd be worth the calculation. Would you agree?

Reply:
Sure, good enough as-is. If someone comes with a source file with more lines we question if there is any value in giving him the right line number since his editor might not be able to load that file anyway :wink:

Reply:
sounds like a plan!